### ISSUE TITLE
GraphQL federation _entities method isn't resolved properly.

### ISSUE THREAD

**Issue by alexpoletaev** [quarkusio/quarkus#30180]
### Describe the bug

GraphQL federation `_entities` method isn't resolved properly. Method execution throws exception:
`java.lang.RuntimeException: can't resolve federated entity type io.smallrye.context.Jdk12CompletableFutureWrapper`.

It seems that it fails to resolve the `_Entity` type here `io.smallrye.graphql.bootstrap.Bootstrap#fetchEntityType`

### Expected behavior

Entity is successfully resolved

### Actual behavior

Exception is thrown

### How to Reproduce?

Let's imagine we have two separated services: `products` and `reviews`. The `reviews` service extends the `Product` entity with some new attributes (`rating` and `reviews`). Here is a reproducer for the reviews service
[quarkus-graphql-issue.zip](https://github.com/quarkusio/quarkus/files/10346373/quarkus-graphql-issue.zip)

- unzip a reproducer
- start service
- go to graphql UI
- execute next query:
```
query representations($rep: [_Any!]!) {
  _entities(representations: $rep) {
    ... on Product {
      rating
    }
  }
}
```
and query variables are:
```
{
  "rep": {
    "__typename": "Product",
    "id": "bfcd3cc0-dc58-4e20-8c4c-a313de3ad8d9"
  }
}
```
- exception is thrown

This method is used by the Apollo server to fetch additional attribute values for the product entity by it's key (`id`) and `__typename`. More information can be found [here](https://www.apollographql.com/docs/federation/entities/). I added a method at the `ProductResource` class to resolve product entity by id and this method is successfully resolved by the graphql module (by the return type), but then the exception is thrown here `io.smallrye.graphql.bootstrap.Bootstrap#fetchEntityType`

### Output of `java -version`

openjdk version "17.0.5" 2022-10-18

### Quarkus version or git rev

2.15.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 7.5.1

### Additional information

It would be great if the highest priority will be assigned to this issue among other graphql issues since there is no workaround found and it not allows to fully use all graphql federation features

**Comment by t1** [comment]
I've looked into this and it works fine without the 'Uni' wrapping. Is that an option?

**Comment by alexpoletaev** [comment]
> I've looked into this and it works fine without the 'Uni' wrapping. Is that an option?

Hm, @t1, could you please provide an example? I've removed the `Uni` wrapping and now I'm getting a little bit another exception:
`java.lang.RuntimeException: can't resolve federated entity type java.util.concurrent.CompletableFuture`

Here is what i've changed:
in `ProductResource` I've refactored the method to:
```
    @Query("productById")
    @Description("Find product by ID")
    public Product getById(UUID id) {
        return service.findById(id);
    }
```
in `ProductService` I've refactored the method to:
```
    public Product findById(UUID id) {
        return Optional.ofNullable(productMap.get(id)).orElseThrow();
    }
```

And, honestly, in real project we use a reactive hibernate and I'm not sure how to remove a Uni wrapper there...

**Comment by t1** [comment]
I did the same thing, but also for the `Review` classes. As it still has problems with the `CompletableFuture`, it could be that.

To resolve the `Uni`, you'll have to do a blocking call, i.e. `uni.await().atMost(Duration.ofMillis(1000));`. But I'm not much into reactive programming, I guess the code just gets too complex for me.

**Comment by alexpoletaev** [comment]
> I did the same thing, but also for the `Review` classes. As it still has problems with the `CompletableFuture`, it could be that.
> 
> To resolve the `Uni`, you'll have to do a blocking call, i.e. `uni.await().atMost(Duration.ofMillis(1000));`. But I'm not much into reactive programming, I guess the code just gets too complex for me.

Replacing all `Uni`s with blocking code doesn't help, I still getting `java.lang.RuntimeException: can't resolve federated entity type java.util.concurrent.CompletableFuture` exception.

I made some debugging/research and the only way I get it work properly is (maybe it will help to fix the bug):

1. Refactor `ProductResource` to
```
    public Product getById(UUID id) {
        return service.findById(id).await().atMost(Duration.ofMillis(500));
    }
```
the `Review` resource and service I left the same as in the reproducer, with `Uni`s.
2. Add config `quarkus.smallrye-graphql.nonblocking.enabled = false`

But, unfortunately, it's not an option for my project, as it makes **all** qraphql requests to be blocking (if I understood correctly) and negates all the Quarkus benefits.

**Comment by t1** [comment]
Hmmmm... this is more complicated than I thought. I had rebuilt your reproducer on WildFly, and it worked there. I'm still investigating.

**Comment by phillip-kruger** [comment]
@t1 we will have to support Uni (and CompletionStage*) out of the box. You can look at how we handle it for normal queries. Uni will have to execute on the correct thread.

**Comment by alexpoletaev** [comment]
> Hmmmm... this is more complicated than I thought. I had rebuilt your reproducer on WildFly, and it worked there. I'm still investigating.

Thank you for checking this! Depending on the method return type, the  `FederationDataFetcher` uses the `QuarkusDefaultDataFetcher` or the `QuarkusUniDataFetcher`. If the latter one is used, the result is wrapped into `CompletableFuture` (or, to be precise, in my case it's a `io.smallrye.context.Jdk12CompletableFutureWrapper`) and the `io.smallrye.graphql.bootstrap.Bootstrap#fetchEntityType` fails to retrieve the return type. I'm not familiar with WildFly, probably it has different implementation for one of those classes.

**Comment by alexpoletaev** [comment]
> Hi @alexpoletaev could you work around this issue? Because I'm also struggeling with the federation thing :(

Hi, as a temporary workaround we replaced federation requests from the frontend with the direct service queries in our project, but this is a very ugly decision. I hope soon I'll find an extra time to figure out what causes the problem and prepare a pull request with the fix

**Comment by steffenthomasEMP** [comment]
Hi @alexpoletaev, I also removed any kind of usage of mutiny. 
1. I do have is that I cannot reproduce on my local machine...with the service running stand-alone, thus, no federation-execution (I always have to push the code to staging-env)
2. I could manage to lead the gateway to route to my service, but then I'm getting following stacktrace:
`ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-2) HTTP Request to /graphql failed,... java.lang.RuntimeException: can't resolve federated entity type io.smallrye.context.CompletableFutureWrapper`

So obviously I have a problem in my schema...maybe a type is missing...but the CompletableFutureWrapper makes me confused...

Actually, we follow the same strategy as you...doing different queries (with no federation-idea)

**Comment by eformat** [comment]
Hi .. i had some success getting this going with an older version of Apollo Gateway (patched to support application/graphql+json)

You can check out the working code here:

https://github.com/eformat/quarkus-graphql-issue

I split it out a bit to make it clearer.

There does not seem to be any issue using Uni's .. i think the CompletableFutureWrapper you are seeing is real - i.e. the subgraph query does not work properly.

HTH
Cheers

**Comment by steffenthomasEMP** [comment]
Hey @eformat, cool, thanks for sharing your findings. 
Actually, I could make it run yesterday evening. So, the GraphQLApi-Resource is out of Uni's and the Service-layer, too. The RestClient (I know this is weird) under the hook is still reactive. The Service then does the most-Duration-blocking-trick. 
Finally the rescue was to add `quarkus.smallrye-graphql.nonblocking.enabled = false` in `META-INF/mircoprofile-config.properties`. AND...I had to define the graphQL-Schema statically on my own, because the smallrye-maven-plugin generates a schema without any infos about the directives like `@key`, which I need to have in federation (even the parameters are set in the plugin). 
The error-message I got then was: `"message": "no query found for Product by [category, id]"`. So I've added that query in the resource BUT not in schema, because that broke the whole gateway to work, earlier. 

So, thank's for your help here. I will definately have a look into your example project. 
Best

**Comment by eformat** [comment]
hi @steffenthomasEMP  .. ahh interesting re: blocking .. erm, try adding:

quarkus.smallrye-graphql.schema-include-scalars=true
quarkus.smallrye-graphql.schema-include-directives=true

that should give you all the directive bits in the schema

**Comment by steffenthomasEMP** [comment]
Hi @eformat, sorry to bother you with all the details of my impl...I'll try to give you only the relevant parts.

Here is my setup:

**META-INF/resources/microprofile-config.properties**:
```txt
# copied from demo project https://github.com/t1/smallrye-graphql-federation-demo.   
smallrye.graphql.errorExtensionFields: code,description,validationErrorType,queryPath
mp.graphql.showErrorMessage: java.lang.RuntimeException
smallrye.graphql.events.enabled: true
smallrye.graphql.schema.includeScalars: true
smallrye.graphql.schema.includeDirectives: true
smallrye.graphql.federation.enabled: true
quarkus.smallrye-graphql.nonblocking.enabled=false
```

**pom.xml**
```xml
<plugin>
    <groupId>io.smallrye</groupId>
    <artifactId>smallrye-graphql-maven-plugin</artifactId>
    <version>2.2.1</version>
    <configuration>
        <destination>${project.build.directory}/generated/schema-resolver.graphql</destination>
        <includeDirectives>true</includeDirectives>
        <includeScalars>true</includeScalars>
        <includeDependencies>true</includeDependencies>
        <includeDependenciesScopes>true</includeDependenciesScopes>
        <includeDependenciesTypes>true</includeDependenciesTypes>
        <includeIntrospectionTypes>true</includeIntrospectionTypes>
    </configuration>
    <executions>
        <execution>
            <goals>
                <goal>generate-schema</goal>
            </goals>
        </execution>
    </executions>
</plugin>
```
**Product.java**
```java
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Extends
@Key(fields = "id")
public class Product {

    @NonNull
    @External
    @Id
    String id;

    @External
    String category;
}
```

**Resource.java**
```java
@GraphQLApi
public class Resource {
    @Query
    public Breadcrumb breadcrumb(@Id int cat, String locale) {
        return the.breadcrumb;
    }

    @Query
    public Product product(@Id String id, String category){
        LOGGER.info("doing federation magic...");
        return Product.builder().id(id).category(category).build();
    }

    public Breadcrumb getBreadcrumb(@Source Product product, String locale, Integer category) {
        return the.breadcrumb.for.this.product;
    }

}
```

**GraphQL-Schema (actual)**
```text
type Product {
  breadcrumb(cat: Int, locale: String): Breadcrumb
  cat: String
  id: ID!
}

"Query root"
type Query {
  breadcrumb(cat: ID!, locale: String): Breadcrumb
  product(category: String, id: ID): Product
}
```

**GraphQL-Schema (expected/needed)**
```text
type Product @extends @key(fields : "id") {
  "The Breadcrumb information"
  breadcrumb(category: Int, locale: String): Breadcrumb @requires(fields : "category")
  category: String @external
  id: ID! @external
}

"Query root"
extend type Query {
  breadcrumb(cat: ID!, locale: String): Breadcrumb
  // the product-query would break gateway-functionality
  //product(category: String, id: ID): Product
}
```

Actually, the last part 'GraphQL-Schema expected/needed' I have statically defined in some place, to be uploaded to schema-registry via pipeline. 
Whenever I try to upload a schema with all the directives and so on, I will get validation errors from gateway because these are centrally managed by the gateway. 
I've also had the issue, when I try to upload a schema with '...type Query {...' the gateway stopped working, because with that the queries from other services had been overwritten...so we are doing hacky scripting to add the 'extend'-keyword to query while uploading the schema. 
For now, I'm fine with the solution, but for the long perspective, I like the way to define the schema code-driven...but with federation in mind it is not working for us, right now. 

And sorry, the discussion is drifting away from the _entities...but anyway, with my setup I came to this discussion-thread

**Comment by ntkoopman** [comment]
`quarkus.smallrye-graphql.nonblocking.enabled=false` also fixes the issue for me.

You can get the subgraph schema from the _sdl query. It will include the directives.
```gql
query { _service { sdl }}
```

**Comment by alexpoletaev** [comment]
@t1 , @phillip-kruger could you please check my [pull request](https://github.com/smallrye/smallrye-graphql/pull/1899)?

**Comment by alexpoletaev** [comment]
@robp94 , nice, no activity for months in this issue, but 2 pull requests almost in the same time :)

**Comment by jmartisk** [comment]
I'm kinda trying both PRs but neither of them seems to solve the
`java.lang.RuntimeException: can't resolve federated entity type io.smallrye.context.CompletableFutureWrapper` for me

**Comment by robp94** [comment]
@jmartisk how did you test this?

**Comment by jmartisk** [comment]
I have two apps that have one federated entity, `Customer`, joined by the ID field.
It's the `acmeshop` and `acmeshop-addons` applications in this directory: https://github.com/jmartisk/mock-artifacts/tree/main/acmeshop
When I run the router as described in https://github.com/jmartisk/mock-artifacts/blob/main/acmeshop/federation.md and try, for example
```
customers {
  name
  password
}
```

Then it can't fetch the password (that field comes from the `acmeshop-addons` app).
But maybe it's wrong on my side, I'm not really familiar with this yet.

**Comment by robp94** [comment]
I will give it a try later today, you do not need a real federation setup to test it.
A query like this:
```
query SomeName($representations: [_Any!]!) {
  _entities(representations: $representations) {
    ... on Resource {
      something {
        __typename
        id
      }
    }
  }
}
```

With variables like this:

```
{
  "representations": [
    {
      "__typename": "Resource",
      "id": "026b309b-0cbc-471b-994c-8ea825b31f1c"
    },
    {
      "__typename": "Resource",
      "id": "04832d89-4aab-4c9d-b595-32f7b7a23548"
    }

  ]
}
```

Should work.

**Comment by robp94** [comment]
@jmartisk I made you a pr.

**Comment by jmartisk** [comment]
Ok so sorry, this is explained now, with @robp94 's PR (https://github.com/smallrye/smallrye-graphql/pull/1898) it seems to work for me after all. I'll have another look tomorrow, need to drop now.

**Comment by alexpoletaev** [comment]
Hm, strange... @jmartisk, I see that you use a blocking request for the customer retrieval in your 'addons' service, but I though we have this issue only with non-blocking requests. And, if I remember correctly, I didn't have such problems in the reproducer from this issue in case of blocking requests. Probably I'm missing something and need to revise the implementation again

**Comment by jmartisk** [comment]
I merged @robp94 's PR as it seems to work for me. It would be nice to have some tests too, so unless somebody volunteers I'll look into that later this week
I think I'll also hurry and make a 2.4.0 release today so that we can make it with this change into Quarkus 3.4

**Comment by robp94** [comment]
@jmartisk if you can point me at the right module, I can write you a federation query/test for a small example later this week. I will although try to add some more features for better performance.

**Comment by jmartisk** [comment]
We have either Arquillian-based tests in https://github.com/smallrye/smallrye-graphql/tree/main/server/integration-tests
or on Quarkus side, there is https://github.com/quarkusio/quarkus/tree/main/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment

(In Quarkus, there is already [GraphQLFederationTest.java](https://github.com/quarkusio/quarkus/blob/main/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/GraphQLFederationTest.java), but it only very superficially looks at directives in the schema, no runtime functionality)

Choose whichever of these modules you prefer

**Comment by alexpoletaev** [comment]
@jmartisk, thank you very much! I will also try to find some time for tests.

**Comment by jmartisk** [comment]
This should be fixed in the upcoming Quarkus 3.4.0
